src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogValuesFormatter.cs

```diff
@@ -263,8 +263,15 @@ private static object FormatArgument(object? value)
             return TryFormatArgumentIfNullOrEnumerable(value, ref stringValue) ? stringValue : value!;
         }
 
-        private static bool TryFormatArgumentIfNullOrEnumerable(object? value, [NotNullWhen(true)] ref object? stringValue)
+        private static bool TryFormatArgumentIfNullOrEnumerable<T>(T? value, [NotNullWhen(true)] ref object? stringValue)
         {
+            // Avoiding boxing of known types
+            if (typeof(T).IsPrimitive || typeof(T).IsEnum)
+            {
+                stringValue = null;
+                return false;
+            }
```

Comment:
What benefit does this block of code provide?  I'm not seeing where boxing would be coming from without it, at least not in optimized code.  For a primitive or an enum, the JIT should effectively get rid of everything else in the method already, since it'll be generically specialized for that value type T and the JIT can tell the value is non-null and not an IEnumerable.

Reply:
> at least not in optimized code.

This is what I was worried about. Maybe I was looking at tier 0 code? I can revert, but while profiling this I was seeing boxing and I assumed it was coming from the interface check (if T is IEnumerable).

Reply:
> Maybe I was looking at tier 0 code?

That's my guess.

> I assumed it was coming from the interface check (if T is IEnumerable).

That shouldn't box in optimized code.  If it does, we should fix the JIT. @EgorBo?

Reply:
It looks like both of you're correct - we optimize this in Tier1 but still allocate in Tier0 (although, should not be a real issue for this case because BCL libs are prejitted, hence, the boxing will be always optimized unless user explicitly defines `DOTNET_ReadyToRun=0`). Filed https://github.com/dotnet/runtime/issues/88987 with a minimal repro

Reply:
The one in VS? Doesn't it disable tiering? by default?

Reply:
I think it turns of R2R because of native symbol resolution or something like that... I can't remember if it also disable tiered compilation. That would end up allocating *A LOT* more right?

Reply:
> That would end up allocating A LOT more right?

No, it would mean everything compiles to tier 1 directly rather than going through tier 0.

Reply:
> That would end up allocating A LOT more right?

No, it would mean everything compiles to tier 1 directly rather than going through tier 0.

Reply:
Got it. Then it's doing both things (turning of R2R and tiered compilation).

Reply:
Hopefully, fixed in https://github.com/dotnet/runtime/pull/88989 so I can probably remove this workaround in that PR (btw, `typeof().IsPrimitive` is not optimized to constant)

Reply:
> typeof().IsPrimitive is not optimized to constant

Good to know. When I was initially looking at it I was wondering if I needed to expand this into a hard coding all of the types...

Reply:
> so I can probably remove this workaround in that PR

I think we should remove it, regardless. We generally don't add code to optimize tier 0.

Reply:
Sure I'll send a PR, I just wanted to understand why I was seeing boxing in the tool 😄.

Reply:
For reference:
<img width="588" alt="image" src="https://github.com/dotnet/runtime/assets/2642209/362b3efe-6e3c-44b8-bce9-ca9d0d15157b">

And:
<img width="1137" alt="image" src="https://github.com/dotnet/runtime/assets/2642209/12ffcfae-5a08-4115-990d-a8f2cf3e45a1">

Reply:
So this doesn't explain why you were seeing boxing from here...

Reply:
https://github.com/dotnet/runtime/pull/89000